README.md
# curved_navigation_bar_with_label

[pub package](https://pub.dartlang.org/packages/curved_navigation_bar_with_label)

This project forked from https://pub.dev/packages/curved_navigation_bar
A Flutter package for easy implementation of curved navigation bar.

![Gif](https://github.com/ARASHz4/curved_navigation_bar_with_label/blob/master/example.gif "Fancy Gif")

### Add dependency

```yaml
dependencies:
  curved_navigation_bar: ^2.1.2 #latest version
```

### Easy to use

```dart
Scaffold(
  bottomNavigationBar: CurvedNavigationBar(
    backgroundColor: Colors.blueAccent,
    items: [
      CurvedNavigationBarItem(icon: Icon(Icons.add, size: 30), label: "Add"),
      CurvedNavigationBarItem(icon: Icon(Icons.list, size: 30), label: "List"),
      CurvedNavigationBarItem(icon: Icon(Icons.compare_arrows, size: 30), label: "Compare"),
    ],
    onTap: (index) {
      //Handle button tap
    },
  ),
  body: Container(color: Colors.blueAccent),
)
```

### Attributes

`items`: List of Widgets\
`index`: index of NavigationBar, can be used to change current index or to set initial index\
`color`: Color of NavigationBar, default Colors.white\
`buttonBackgroundColor`: background color of floating button, default same as color attribute\
`backgroundColor`: Color of NavigationBar's background, default Colors.blueAccent\
`onTap`: Function handling taps on items\
`animationCurve`: Curves interpolating button change animation, default Curves.easeOutCubic\
`animationDuration`: Duration of button change animation, default Duration(milliseconds: 600)\
`height`: Height of NavigationBar, min 0.0, max 75.0\
`letIndexChange`: Function which takes page index as argument and returns bool. If function returns
false then page is not changed on button tap. It returns true by default\

### Change page programmatically

```dart
 //State class
int _page = 0;
GlobalKey<CurvedNavigationBarState> _bottomNavigationKey = GlobalKey();

@override
Widget build(BuildContext context) {
  return Scaffold(
    bottomNavigationBar: CurvedNavigationBar(
      key: _bottomNavigationKey,
      items: [
        CurvedNavigationBarItem(icon: Icon(Icons.add, size: 30), label: "Add"),
        CurvedNavigationBarItem(icon: Icon(Icons.list, size: 30), label: "List"),
        CurvedNavigationBarItem(icon: Icon(Icons.compare_arrows, size: 30), label: "Compare"),
      ],
      onTap: (index) {
        setState(() {
          _page = index;
        });
      },
    ),
    body: Container(
      color: Colors.blueAccent,
      child: Center(
        child: Column(
          children: <Widget>[
            Text(_page.toString(), textScaleFactor: 10.0),
            ElevatedButton(
              child: Text('Go To Page of index 1'),
              onPressed: () {
                //Page change using state does the same as clicking index 1 navigation button
                final CurvedNavigationBarState? navBarState =
                    _bottomNavigationKey.currentState;
                navBarState?.setPage(1);
              },
            )
          ],
        ),
      ),
    ),
  );
}
```

CHANGELOG.md
## 2.1.4
- fixed bugs

## 2.1.2
- added label

## 2.1.0
- added label
- fixed bugs


lib/nav_custom_painter.dart
import 'package:flutter/material.dart';

class NavCustomPainter extends CustomPainter {
  final s = 0.2;

  late double loc;
  late double bottom;
  Color color;
  TextDirection textDirection;

  NavCustomPainter({
    required double startingLoc,
    required int itemsLength,
    required this.color,
    required this.textDirection,
  }) {
    final span = 1.0 / itemsLength;
    final l = startingLoc + (span - s) / 2;
    loc = textDirection == TextDirection.rtl ? 0.8 - l : l;
    bottom = 0.64;
  }

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = color
      ..style = PaintingStyle.fill;

    final path = Path()
      ..moveTo(0, 0)
      ..lineTo(size.width * (loc - 0.05), 0)
      ..cubicTo(
        size.width * (loc + s * 0.2), // topX
        size.height * 0.05, // topY
        size.width * loc, // bottomX
        size.height * bottom, // bottomY
        size.width * (loc + s * 0.5), // centerX
        size.height * bottom, // centerY
      )
      ..cubicTo(
        size.width * (loc + s), // bottomX
        size.height * bottom, // bottomY
        size.width * (loc + s * 0.8), // topX
        size.height * 0.05, // topY
        size.width * (loc + s + 0.05),
        0,
      )
      ..lineTo(size.width, 0)
      ..lineTo(size.width, size.height)
      ..lineTo(0, size.height)
      ..close();
    canvas.drawPath(path, paint);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return this != oldDelegate;
  }
}

lib/curved_navigation_bar.dart
import 'package:curved_navigation_bar_with_label/nav_custom_painter.dart';
import 'package:flutter/material.dart';

typedef LetIndexPage = bool Function(int value);

int _currentIndex = 0;

class CurvedNavigationBar extends StatefulWidget {
  final List<CurvedNavigationBarItem> items;
  final int index;
  final Color? color;
  final Color backgroundColor;
  final Color? buttonBackgroundColor;
  final Color? buttonLabelColor;
  final ValueChanged<int>? onTap;
  final LetIndexPage letIndexChange;
  final Curve animationCurve;
  final Duration animationDuration;
  final double height;

  CurvedNavigationBar({
    Key? key,
    required this.items,
    this.index = 0,
    this.color,
    this.backgroundColor = Colors.blueAccent,
    this.buttonBackgroundColor,
    this.buttonLabelColor,
    this.onTap,
    LetIndexPage? letIndexChange,
    this.animationCurve = Curves.easeOut,
    this.animationDuration = const Duration(milliseconds: 600),
    this.height = 68.0,
  })  : letIndexChange = letIndexChange ?? ((_) => true),
        assert(items.isNotEmpty),
        assert(0 <= index && index < items.length),
        assert(0 <= height && height <= height),
        super(key: key);

  @override
  CurvedNavigationBarState createState() => CurvedNavigationBarState();
}

class CurvedNavigationBarState extends State<CurvedNavigationBar>
    with SingleTickerProviderStateMixin {
  late double _startingPos;
  int _endingIndex = 0;
  late double _pos;
  double _buttonHide = 0;
  late Widget _icon;
  late String _label;
  late AnimationController _animationController;
  late int _length;

  @override
  void initState() {
    super.initState();

    _icon = widget.items[widget.index].icon;
    _label = widget.items[widget.index].label;
    _length = widget.items.length;
    _pos = widget.index / _length;
    _startingPos = widget.index / _length;

    _animationController = AnimationController(vsync: this, value: _pos);
    _animationController.addListener(() {
      setState(() {
        _pos = _animationController.value;
        final endingPos = _endingIndex / widget.items.length;
        final middle = (endingPos + _startingPos) / 2;
        if ((endingPos - _pos).abs() < (_startingPos - _pos).abs()) {
          _icon = widget.items[_endingIndex].icon;
          _label = widget.items[_endingIndex].label;
        }
        _buttonHide =
            (1 - ((middle - _pos) / (_startingPos - middle)).abs()).abs();
      });
    });
  }

  @override
  void didUpdateWidget(CurvedNavigationBar oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.index != widget.index) {
      final newPosition = widget.index / _length;
      _startingPos = _pos;
      _endingIndex = widget.index;
      _animationController.animateTo(newPosition,
          duration: widget.animationDuration, curve: widget.animationCurve);
    }
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    _label = widget.items[_endingIndex].label;
    _length = widget.items.length;

    Size size = Size(
        MediaQuery.of(context).size.width -
            (MediaQuery.of(context).padding.left +
                MediaQuery.of(context).padding.right),
        MediaQuery.of(context).size.height);

    return SafeArea(
      child: Material(
        color: widget.backgroundColor,
        child: Stack(
          clipBehavior: Clip.none,
          alignment: Alignment.bottomCenter,
          children: [
            Positioned(
              bottom: -40 - (75.0 - widget.height),
              left: Directionality.of(context) == TextDirection.rtl
                  ? null
                  : _pos * size.width,
              right: Directionality.of(context) == TextDirection.rtl
                  ? _pos * size.width
                  : null,
              width: size.width / _length,
              child: Center(
                child: Transform.translate(
                  offset: Offset(
                    0,
                    -(1 - _buttonHide) * 80,
                  ),
                  child: Material(
                    color: widget.buttonBackgroundColor ??
                        Theme.of(context).cardColor,
                    type: MaterialType.circle,
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: _icon,
                    ),
                  ),
                ),
              ),
            ),
            CustomPaint(
              painter: NavCustomPainter(
                startingLoc: _pos,
                itemsLength: _length,
                color: Theme.of(context).scaffoldBackgroundColor,
                textDirection: Directionality.of(context),
              ),
              child: Container(
                height: widget.height,
              ),
            ),
            Positioned(
              left: Directionality.of(context) == TextDirection.rtl
                  ? null
                  : _pos * size.width,
              right: Directionality.of(context) == TextDirection.rtl
                  ? _pos * size.width
                  : null,
              width: size.width / _length,
              child: Center(
                child: Padding(
                  padding: const EdgeInsets.only(bottom: 2),
                  child: Text(
                    _label,
                    style: TextStyle(
                        color:
                            widget.buttonLabelColor ?? widget.backgroundColor,
                        fontWeight: FontWeight.bold),
                    maxLines: 1,
                  ),
                ),
              ),
            ),
            SizedBox(
              height: widget.height,
              child: Row(
                children: widget.items.map(
                  (item) {
                    return NavButton(
                      onTap: _buttonTap,
                      position: _pos,
                      length: _length,
                      index: widget.items.indexOf(item),
                      height: widget.height,
                      child: Center(
                        child: Stack(
                          children: [
                            Padding(
                              padding: const EdgeInsets.only(top: 12),
                              child: Align(
                                alignment: Alignment.topCenter,
                                child: item.icon,
                              ),
                            ),
                            Padding(
                              padding: const EdgeInsets.only(bottom: 4),
                              child: Align(
                                alignment: Alignment.bottomCenter,
                                child: Text(
                                  item.label,
                                  maxLines: 1,
                                ),
                              ),
                            ),
                          ],
                        ),
                      ),
                    );
                  },
                ).toList(),
              ),
            ),
          ],
        ),
      ),
    );
  }

  setPage(int index) {
    _buttonTap(index);
  }

  _buttonTap(int index) {
    if (widget.onTap != null) {
      widget.onTap!(index);
    }

    if (_endingIndex != index) {
      if (!widget.letIndexChange(index)) {
        return;
      }

      final newPosition = index / _length;
      setState(() {
        _startingPos = _pos;
        _endingIndex = index;
        _currentIndex = index;
        _animationController.animateTo(newPosition,
            duration: widget.animationDuration, curve: widget.animationCurve);
      });
    }
  }
}

class CurvedNavigationBarItem {
  final Widget icon;
  final String label;

  CurvedNavigationBarItem({required this.icon, required this.label});
}

class NavButton extends StatelessWidget {
  final double position;
  final int length;
  final int index;
  final ValueChanged<int> onTap;
  final Widget child;
  final double height;

  const NavButton({
    Key? key,
    required this.onTap,
    required this.position,
    required this.length,
    required this.index,
    required this.child,
    required this.height,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final desiredPosition = 1.0 / length * index;
    final difference = (position - desiredPosition).abs();
    final verticalAlignment = 1 - length * difference;
    final opacity = length * difference;

    return Expanded(
      child: Material(
        color: Colors.transparent,
        child: _currentIndex != index
            ? InkWell(
                customBorder:
                    const CircleBorder(), //const ContinuousRectangleBorder(borderRadius: BorderRadius.all(Radius.circular(50))),
                onTap: () {
                  onTap(index);
                },
                child: SizedBox(
                  height: height,
                  child: Transform.translate(
                    offset: Offset(0,
                        difference < 1.0 / length ? verticalAlignment * 40 : 0),
                    child: Opacity(
                        opacity:
                            difference < 1.0 / length * 0.99 ? opacity : 1.0,
                        child: child),
                  ),
                ),
              )
            : GestureDetector(
                behavior: HitTestBehavior.translucent,
                onTap: () {
                  onTap(index);
                },
                child: SizedBox(
                  height: height,
                  child: Transform.translate(
                    offset: Offset(0,
                        difference < 1.0 / length ? verticalAlignment * 40 : 0),
                    child: Opacity(
                        opacity:
                            difference < 1.0 / length * 0.99 ? opacity : 1.0,
                        child: child),
                  ),
                ),
              ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:curved_navigation_bar_with_label/curved_navigation_bar.dart';
import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(home: BottomNavBar()));

class BottomNavBar extends StatefulWidget {
  @override
  _BottomNavBarState createState() => _BottomNavBarState();
}

class _BottomNavBarState extends State<BottomNavBar> {
  int _page = 0;
  GlobalKey<CurvedNavigationBarState> _bottomNavigationKey = GlobalKey();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: CurvedNavigationBar(
        key: _bottomNavigationKey,
        index: 0,
        height: 76.0,
        items: [
          CurvedNavigationBarItem(icon: Icon(Icons.add, size: 30), label: "Add"),
          CurvedNavigationBarItem(icon: Icon(Icons.list, size: 30), label: "List"),
          CurvedNavigationBarItem(icon: Icon(Icons.compare_arrows, size: 30), label: "Compare"),
          CurvedNavigationBarItem(icon: Icon(Icons.call_split, size: 30), label: "Split"),
          CurvedNavigationBarItem(icon: Icon(Icons.perm_identity, size: 30), label: "User"),
        ],
        backgroundColor: Colors.blueAccent,
        animationCurve: Curves.easeInOut,
        animationDuration: Duration(milliseconds: 600),
        onTap: (index) {
          setState(() {
            _page = index;
          });
        },
        letIndexChange: (index) => true,
      ),
      body: Container(
        color: Colors.blueAccent,
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text(_page.toString(), textScaleFactor: 10.0),
              ElevatedButton(
                child: Text('Go To Page of index 1'),
                onPressed: () {
                  final CurvedNavigationBarState? navBarState = _bottomNavigationKey.currentState;
                  navBarState?.setPage(1);
                },
              )
            ],
          ),
        ),
      ),
    );
  }
}
